majel/place.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+#include <boost/variant.hpp>
+#include <iostream>
+
+namespace majel {
+
+struct CpuPlace {
+  CpuPlace() {}  // WORKAROUND: for some reason, omitting this constructor
+                 // causes errors with boost 1.59 and OSX
+  // needed for variant equality comparison
+  inline bool operator==(const CpuPlace&) const { return true; }
+
+  inline bool operator!=(const CpuPlace&) const { return false; }
+};
+
+struct GpuPlace {
+  GpuPlace(int d) : device(d) {}
+
+  // needed for variant equality comparison
+  inline bool operator==(const GpuPlace& o) const { return device == o.device; }
+
+  inline bool operator!=(const GpuPlace& o) const { return !(*this == o); }
```

Comment:
@wangkuiyi @helinwang I think it's correct. Because `*this == o` will trigger `inline bool operator==(const GpuPlace& o) `